pkg/daemon/cleanup/cleanup.go

```diff
@@ -0,0 +1,97 @@
+//go:build windows
+
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cleanup
+
+import (
+	"context"
+	"net"
+
+	"github.com/pkg/errors"
+	core "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/openshift/windows-machine-config-operator/pkg/daemon/controller"
+	"github.com/openshift/windows-machine-config-operator/pkg/daemon/manager"
+	"github.com/openshift/windows-machine-config-operator/pkg/nodeconfig"
+	"github.com/openshift/windows-machine-config-operator/pkg/servicescm"
+)
+
+// Deconfigure removes an instance from the cluster. If the parameter flag is true, the node object is not deleted.
+// If we are able to get the services ConfigMap tied to the desired version, all services defined in it are cleaned up.
+// Otherwise, perform cleanup based on a combination of the OpenShift managed tag and the latest services ConfigMap.
+func Deconfigure(c client.Client, netInterfaceAddresses []net.Addr, svcMgr manager.Manager, preserveNode bool) error {
+	var node *core.Node
+	ctx := context.TODO()
+	cm := &core.ConfigMap{}
+	// Fetch the CM of the desired version
+	err := func() error {
+		var err error
+		node, err = controller.GetAssociatedNode(c, netInterfaceAddresses)
+		if err != nil {
+			return err
+		}
+		desiredVersion, present := node.Annotations[nodeconfig.DesiredVersionAnnotation]
+		if !present {
+			return errors.Wrapf(err, "node missing desired version annotation")
+		}
+		return c.Get(ctx,
+			client.ObjectKey{Namespace: controller.WMCONamespace, Name: servicescm.NamePrefix + desiredVersion}, cm)
+	}()
+	if err != nil {
+		// TODO: best effort cleanup of all OpenShift managed services and those in the latest services ConfigMap
+		return err
+	} else {
+		// Happy path, no issues getting the desired services ConfigMap
+		if err = removeCMServices(c, ctx, svcMgr, cm); err != nil {
+			return err
+		}
+	}
+	// Delete node if needed
+	if preserveNode || node == nil {
+		return nil
+	}
+	klog.Infof("deleting node %s", node.GetName())
+	return c.Delete(ctx, node)
+}
+
+// removeCMServices removes all the services defined in the given ConfigMap from this instance.
+func removeCMServices(c client.Client, ctx context.Context, svcMgr manager.Manager, cm *core.ConfigMap) error {
+	cmData, err := servicescm.Parse(cm.Data)
```

Comment:
If `cm.Data` is what the function needs, then we should only be passing that in.